pallets/xcm-weight-trader/src/lib.rs

```diff
@@ -422,32 +422,27 @@ impl<T: crate::Config> WeightTrader for Trader<T> {
 			id: XcmAssetId(location),
 		}) = self.1.take()
 		{
-			if actual_weight == self.0 {
-				self.1 = Some(Asset {
-					fun: Fungibility::Fungible(initial_amount),
-					id: XcmAssetId(location),
-				});
-				None
-			} else {
-				let weight = actual_weight.min(self.0);
-				let amount: u128 =
-					Self::compute_amount_to_charge(&weight, &location).unwrap_or(u128::MAX);
-				let final_amount = amount.min(initial_amount);
-				let amount_to_refund = initial_amount.saturating_sub(final_amount);
-				self.0 -= weight;
-				self.1 = Some(Asset {
-					fun: Fungibility::Fungible(final_amount),
-					id: XcmAssetId(location.clone()),
-				});
-				log::trace!(
-					target: "xcm-weight-trader",
-					"refund_weight amount to refund: {:?}",
-					amount_to_refund
-				);
+			let weight = actual_weight.min(self.0);
+			let amount: u128 =
+				Self::compute_amount_to_charge(&weight, &location).unwrap_or(u128::MAX);
+			let amount_to_refund = initial_amount.saturating_sub(amount);
+			self.0 -= weight;
+			self.1 = Some(Asset {
+				fun: Fungibility::Fungible(amount_to_refund),
+				id: XcmAssetId(location.clone()),
+			});
+			log::trace!(
+				target: "xcm-weight-trader",
+				"refund_weight amount to refund: {:?}",
+				amount_to_refund
+			);
+			if amount > 0 {
```

Comment:
Make the implementation the same as https://paritytech.github.io/polkadot-sdk/master/staging_xcm_builder/struct.UsingComponents.html 

This resulted in fixing refund_surplus instruction benchmark test which was failing before.

Reply:
Just replace `amount` by `amount_to_refund` lines 439 and 441.
The  `amount` is not what we refund, it's what the user actually have to pay